### PR TITLE
npc_control: Advanced NPC disabling

### PR DESCRIPTION
**Pull request by lnvoker**
Main changes:
1) Allows specific NPCs to spawn even at high server load.
2) Replaces basic FLHook NPC disabling. Mission and .aicreate NPCs still spawn at any server load.

### CODE REVIEW

File: Plugins/Public/npc_control/Main.cpp [hunk 1]
```diff
@@ -852,6 +874,56 @@ bool ExecuteCommandString_Callback(CCmds* cmds, const wstring &wscCmd)
...
+
+		// Distinguish player ships from NPC ships to avoid unpleasant cases.
+		if (HkGetClientIDByShip(shipInfo.iSpaceID))
+			return true;
+
+		pub::AI::Personality pers;
+		CShip* ship = reinterpret_cast<CShip*>(CObject::Find(shipInfo.iSpaceID, CObject::CSHIP_OBJECT));
+		ship->get_behavior_interface()->get_personality(pers);
+
+		// Distinguish mission NPCs from regular to avoid unpleasant cases.
```
Comment (dsyalex) [inline Plugins/Public/npc_control/Main.cpp]:
```suggestion
		// Distinguish mission NPCs from regular - we want to disable random spawns, never mission NPCs
```

File: Plugins/Public/npc_control/Main.cpp [hunk 2]
```diff
@@ -852,6 +874,56 @@ bool ExecuteCommandString_Callback(CCmds* cmds, const wstring &wscCmd)
...
+		if (pers.Job.field_targeting == 2)
+			return true;
+
+		// Distinguish FLHook NPCs from regular to avoid unpleasant cases.
```
Comment (dsyalex) [inline Plugins/Public/npc_control/Main.cpp]:
```suggestion
		// Distinguish FLHook NPCs from regular - we want to disable random spawns, never FLHook NPCs
```

File: Plugins/Public/npc_control/Main.cpp [hunk 3]
```diff
@@ -852,6 +874,56 @@ bool ExecuteCommandString_Callback(CCmds* cmds, const wstring &wscCmd)
...
+{
+	returncode = DEFAULT_RETURNCODE;
+
+	if (g_iServerLoad > ServerLoad_DisableSpawn)
+	{
+		// Tell FLHook that we disable NPCs. Needed for .serverinfo command.
+		g_bNPCDisabled = true;
+
+		// Distinguish player ships from NPC ships to avoid unpleasant cases.
+		if (HkGetClientIDByShip(shipInfo.iSpaceID))
```
Comment (dsyalex) [inline Plugins/Public/npc_control/Main.cpp]: Can we use ship->is_player() for this?

File: Plugins/Public/npc_control/Main.cpp [hunk 4]
```diff
@@ -852,6 +874,56 @@ bool ExecuteCommandString_Callback(CCmds* cmds, const wstring &wscCmd)
...
+		returncode = SKIPPLUGINS_NOFUNCTIONCALL; // Prevent client-side spawning.
+	}
+	else
+	{
+		// Tell FLHook that we enable NPCs. Needed for .serverinfo command.
+		g_bNPCDisabled = false;
+	}
+
+
+	return false;
```
Comment (dsyalex) [inline Plugins/Public/npc_control/Main.cpp]: don't we want to return true here as well?

File: Plugins/Public/npc_control/Main.cpp [hunk 5]
```diff
@@ -852,6 +874,56 @@ bool ExecuteCommandString_Callback(CCmds* cmds, const wstring &wscCmd)
...
+			return true;
+
+		// If the ship is not present in exception list, prevent it from being created.
+		pub::SpaceObj::Destroy(shipInfo.iSpaceID, VANISH); // Prevent server-side spawning.
+		returncode = SKIPPLUGINS_NOFUNCTIONCALL; // Prevent client-side spawning.
+	}
+	else
+	{
+		// Tell FLHook that we enable NPCs. Needed for .serverinfo command.
+		g_bNPCDisabled = false;
```
Comment (dsyalex) [inline Plugins/Public/npc_control/Main.cpp]: what about the current threshold stuff?

File: Plugins/Public/npc_control/Main.cpp (hunk 3 above)
Comment (lnvoker) [reply Plugins/Public/npc_control/Main.cpp]: We can, but what does this give us.

**Pull request merged**